runtime/pkg/rduckdb/motherduck.go

```diff
@@ -0,0 +1,458 @@
+	}
+
+	spanOptions := otelsql.SpanOptions{
+		SpanFilter: func(ctx context.Context, method otelsql.Method, query string, args []driver.NamedValue) bool {
+			// log all queries except create secret which can contain sensitive data
+			return !createSecretRegex.MatchString(query)
+		},
+	}
+	db := sqlx.NewDb(otelsql.OpenDB(connector, otelsql.WithSpanOptions(spanOptions)), "duckdb")
+	defer func() {
+		if dbErr != nil && db != nil {
+			_ = db.Close()
+		}
+	}()
+
+	// run init queries
+	for _, qry := range opts.DBInitQueries {
+		_, err := db.ExecContext(ctx, qry)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// install and load motherduck database
+	_, err = db.ExecContext(ctx, "INSTALL 'motherduck'; LOAD 'motherduck';")
+	if err != nil {
+		return nil, fmt.Errorf("unable to install motherduck extension: %w", err)
+	}
+	// set the token
+	_, err = db.ExecContext(ctx, fmt.Sprintf("SET motherduck_token=%s;", safeSQLString(opts.Token)))
+	if err != nil && !strings.Contains(err.Error(), "can only be set during initialization") {
+		// ignore `can only be set during initialization` error
+		// it is also returned when database is reopened
+		// looks like the driver or the extension is caching some state
+		return nil, fmt.Errorf("unable to set motherduck token: %w", err)
+	}
+	// attach the database
+	_, err = db.ExecContext(ctx, fmt.Sprintf("ATTACH %s", safeSQLString(opts.dsn())))
+	if err != nil && !strings.Contains(err.Error(), "already exists") {
+		return nil, fmt.Errorf("unable to attach motherduck database: %w", err)
+	}
+
+	err = otelsql.RegisterDBStatsMetrics(db.DB, otelsql.WithAttributes(opts.OtelAttributes...))
+	if err != nil {
+		return nil, fmt.Errorf("registering db stats metrics: %w", err)
+	}
+	return &motherduck{
+		db:       db,
+		logger:   opts.Logger,
+		writeSem: semaphore.NewWeighted(1),
+		opts:     opts,
+	}, nil
+}
+
+// AcquireReadConnection implements DB. In practice this does not enforce that the returned connection is read-only.
+func (m *motherduck) AcquireReadConnection(ctx context.Context) (conn *sqlx.Conn, release func() error, err error) {
+	conn, err = m.acquireConn(ctx)
+	if err != nil {
+		return nil, nil, fmt.Errorf("acquire read connection failed: %w", err)
+	}
+	return conn, conn.Close, nil
+}
+
+// Close implements DB.
+func (m *motherduck) Close() error {
+	return m.db.Close()
+}
+
+// CreateTableAsSelect implements DB.
+func (m *motherduck) CreateTableAsSelect(ctx context.Context, name, query string, opts *CreateTableOptions) (res *TableWriteMetrics, err error) {
+	ctx, span := tracer.Start(ctx, "CreateTableAsSelect", trace.WithAttributes(
+		attribute.String("name", name),
+		attribute.String("query", query),
+		attribute.Bool("view", opts.View),
+	))
+	defer func() {
+		if err != nil {
+			span.SetStatus(codes.Error, err.Error())
+		}
+		span.End()
+	}()
+
+	m.logger.Debug("create: create motherduck table", zap.String("name", name), zap.Bool("view", opts.View), observability.ZapCtx(ctx))
```

Comment:
nit: debug logs should prefix with the file/package name, not function name (to make them easier to find in a log file with debug logs from many different packages). E.g. `motherduck: creating table`.